vehicle/src/Vehicle/Backend/LossFunction/Compile.hs

```diff
@@ -217,6 +213,197 @@ compileQuant :: V.Quantifier -> Quantifier
 compileQuant V.Forall = All
 compileQuant V.Exists = Any
 
+data Translation = Translation
+    { compileAnd          :: LExpr -> LExpr -> LExpr,
+      compileOr           :: LExpr -> LExpr -> LExpr,
+      compileNot          :: LExpr -> LExpr,
+      compileImplication  :: LExpr -> LExpr -> LExpr,
+
+      compileLe           :: LExpr -> LExpr -> LExpr,
+      compileLt           :: LExpr -> LExpr -> LExpr,
+      compileGe           :: LExpr -> LExpr -> LExpr,
+      compileGt           :: LExpr -> LExpr -> LExpr,
+      compileEq           :: LExpr -> LExpr -> LExpr,
+      compileNeq          :: LExpr -> LExpr -> LExpr,
+
+      compileBool         :: Bool -> Double
+    }
+
+---------------------------------------------------------------------------------------
+-- different differentiable logic translations (options were listed before. DL2 is the default translation)
+
+dl2Translation :: Translation
+dl2Translation = Translation
+   { --double check implication, do negation properly
+     compileAnd = Addition,
+     compileOr  = Multiplication,
+     compileNot = id, --this should be normalised out and pushed to the innermost level of comparisons by now
```

Comment:
This is the only thing that worries me in a major way. If that particular normalisation is not done here, then that means that the negation is normalised out for every logic, not just DL2 right? We don't want that?

However, that may mean that the translation will have to accept terms of type `CheckedExpr` rather than `LExpr` which sucks...

Reply:
That's one I worried about as well, but came to the same conclusion and didn't want to start passing `CheckedExpr ` into translations... I'll fix the other issues for now and think about this. In theory normalising the nagtion shouldn't harm the other translations but then they don't need it so it's adding unnecessary complications... 

The only other option I can see is doing some sort of case split on `NotExpr` to have it have different behaviour depending on translation (`compileNot` for all the other ones and the additional negation normalisation for DL2 only?)

Reply:
Hmm okay, so I think the answer is to have the type of the `compileNot` field in the `Translation` record to have the type `Maybe (LExpr -> LExpr -> LExpr)` rather than `LExpr -> LExpr -> LExpr`. A `Nothing` value then indicates that it doesn't exist, and that the `not` should be pushed through. Then to use it, you have to case split on `compileNot` first.

Thoughts?